src/lib.rs

```diff
@@ -726,30 +736,40 @@ impl Regex {
     fn new_options(options: RegexOptions) -> Result<Regex> {
         let raw_tree = Expr::parse_tree_with_flags(&options.pattern, options.compute_flags())?;
 
-        // wrapper to search for re at arbitrary start position,
-        // and to capture the match bounds
+        // after wrapping to:
+        // - search for re at arbitrary start position
+        // - and capture the match bounds
+        // try to optimize the expression tree
         let tree = wrap_tree(raw_tree);
-
+        let (tree, requires_capture_group_fixup) = optimize(tree)?;
         let info = analyze(&tree)?;
 
-        let inner_info = &info.children[1].children[0]; // references inner expr
-        if !inner_info.hard {
+        if !info.hard {
             // easy case, wrap regex
 
             // we do our own to_str because escapes are different
+            // NOTE: there is a good opportunity here to use Hir to avoid regex-automata re-parsing it
             let mut re_cooked = String::new();
             // same as raw_tree.expr above, but it was moved, so traverse to find it
-            let raw_e = match tree.expr {
-                Expr::Concat(ref v) => match v[1] {
-                    Expr::Group(ref child) => child,
+            if !requires_capture_group_fixup {
+                let raw_e = match tree.expr {
+                    Expr::Concat(ref v) => match v[1] {
+                        Expr::Group(ref child) => child,
+                        _ => unreachable!(),
+                    },
                     _ => unreachable!(),
-                },
-                _ => unreachable!(),
+                };
+                raw_e.to_str(&mut re_cooked, 0);
+            } else {
+                tree.expr.to_str(&mut re_cooked, 0);
```

Comment:
So in the case where we optimized, we're now asking the regex crate to match on`(?s:.)*?(pattern)`, instead of just `pattern`.. Does that change anything about how the matching works?

Basically I'm wondering whether the `(?s:.)*?` in the beginning has any bad effects (and we don't really need it for delegating, right?)